hoodie-common/src/main/java/com/uber/hoodie/common/util/buffer/BufferedIteratorExecutor.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2018 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package com.uber.hoodie.common.util.buffer;
+
+import com.uber.hoodie.exception.HoodieException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.function.Function;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+/**
+ * Executor for a BufferedIterator operation. This class takes as input the size limit, buffer producer and transformer
+ * and exposes an API for custom processing of buffered entries
+ */
+public class BufferedIteratorExecutor<I, O, E> {
+
+  private static Logger logger = LogManager.getLogger(BufferedIteratorExecutor.class);
+
+  // Executor service used for launching writer thread.
+  private final ExecutorService executorService;
+  // Used for buffering records which is controlled by HoodieWriteConfig#WRITE_BUFFER_LIMIT_BYTES.
+  private final MemoryBoundedBuffer<I, O> buffer;
+  // Buffer Producer
+  private final BufferProducer<I> producer;
+
+  public BufferedIteratorExecutor(final long bufferLimitInBytes, BufferProducer<I> producer,
+      final Function<I, O> bufferedIteratorTransform) {
+    this.producer = producer;
+    // Ensure single thread for each producer thread and one for consumer
+    this.executorService = Executors.newFixedThreadPool(producer.getNumThreadsNeeded() + 1);
+    this.buffer = new MemoryBoundedBuffer<>(bufferLimitInBytes, bufferedIteratorTransform);
+  }
+
+  /**
+   * Callback to implement environment specific behavior
+   */
+  public void onStartingAsyncThread() {
```

Comment:
can we rename this method to be more clearer about purpose? what's the async thread being referred to here? Producer thread?

Reply:
discussed with balaji f2f on adjustments here